packages/flutter/lib/src/widgets/heroes.dart

```diff
@@ -261,6 +262,8 @@ class _HeroState extends State<Hero> {
 
   @override
   Widget build(BuildContext context) {
+    assert(widget.child is! Hero);
```

Comment:
This does not fully fix https://github.com/flutter/flutter/issues/1398. What if

```dart
child = Container(child: Hero(...))
```

Reply:
Good point!

Reply:
This would work if we did something like:

```dart
assert(() {
  AbstractNode parent = widget.parent;
  while (parent != null) {
    if (parent is Hero) {
      return false;
    }
    parent = parent.parent;
  }
  return true;
}());
```

This isn't as cheap, but should be ok.

Reply:
Actually that wouldn't work - you'd need to test this on the RenderObject, not the Widget, and we don't currently have a specific RenderObject for heroes.  Checking the child definitely isn't sufficient though.

Reply:
@dnfield thanks for your input! 

I actually thought to do a backward traverse of the component tree but thought that it wouldn't be performant enough ... 

I will try to come up with a better solution than what I have. It is as you both stated not sufficient.

Reply:
Ok, so the way you could actually do this is from the build context.  

```dart
@override
void didChangeDependencies() {
  super.didChangeDependencies();
  asert(() {
    return context.ancestorWidgetOfExactType(Hero) == null;
  }());
}
```

Reply:
On further investigation, this would be safe to call from `build`.

Reply:
TIL! I totally missed `ancestorWidgetOfExactType` in `BuildContext` ...

Thank @dnfield it seems you did all the heavy lifting on this 💪 🙏